src/classes/Bot.ts

```diff
@@ -721,329 +717,249 @@ export default class Bot {
         this.addListener(this.client, 'webSession', this.onWebSession.bind(this), false);
         this.addListener(this.client, 'steamGuard', this.onSteamGuard.bind(this), false);
         this.addListener(this.client, 'loginKey', this.handler.onLoginKey.bind(this.handler), false);
-        this.addListener(this.client, 'error', this.onError.bind(this), false);
+        this.addAsyncListener(this.client, 'error', this.onError.bind(this), false);
 
         this.addListener(this.community, 'sessionExpired', this.onSessionExpired.bind(this), false);
-        this.addListener(this.community, 'confKeyNeeded', this.onConfKeyNeeded.bind(this), false);
+        this.addAsyncListener(this.community, 'confKeyNeeded', this.onConfKeyNeeded.bind(this), false);
 
         this.addListener(this.manager, 'pollData', this.handler.onPollData.bind(this.handler), false);
         this.addListener(this.manager, 'newOffer', this.trades.onNewOffer.bind(this.trades), true);
         this.addListener(this.manager, 'sentOfferChanged', this.trades.onOfferChanged.bind(this.trades), true);
         this.addListener(this.manager, 'receivedOfferChanged', this.trades.onOfferChanged.bind(this.trades), true);
         this.addListener(this.manager, 'offerList', this.trades.onOfferList.bind(this.trades), true);
 
-        return new Promise((resolve, reject) => {
-            async.eachSeries(
-                [
-                    (callback): void => {
-                        log.debug('Calling onRun');
-                        void this.handler.onRun().asCallback((err, v) => {
-                            if (err) {
-                                /* eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-call */
-                                return callback(err);
-                            }
+        let promise = Promise.resolve();
 
-                            data = v;
+        const checkIfStopping = () => {
+            if (this.botManager.isStopping) throw new Error('Bot stopping');
+        };
```

Comment:
This function is called in between each promise in the chain and should handle any shutdowns/signals by throwing an error, skipping any further promises added using `.then(func)`. I'm not sure why it's not working.